stream-chat-android-client/src/main/java/io/getstream/chat/android/client/extensions/internal/Channel.kt

```diff
@@ -48,28 +48,37 @@ public val Channel.lastMessage: Message?
 
 @InternalStreamChatApi
 public fun Channel.updateLastMessage(
+    receivedEventDate: Date,
     message: Message,
     currentUserId: String,
 ): Channel {
     val createdAt = message.createdAt ?: message.createdLocallyAt
-    val messageCreatedAt =
-        checkNotNull(createdAt) { "created at cant be null, be sure to set message.createdAt" }
+    checkNotNull(createdAt) { "created at cant be null, be sure to set message.createdAt" }
+
+    val newMessages = (
+        messages
+            .associateBy { it.id } + (message.id to message)
+        )
+        .values
+        .sortedBy { it.createdAt ?: it.createdLocallyAt }
 
-    val updateNeeded = message.id == lastMessage?.id
-    val newLastMessage = lastMessageAt == null || messageCreatedAt.after(lastMessageAt)
     val newReads = read.map { read ->
         read.takeUnless { it.user.id == currentUserId }
             ?: read.copy(
-                lastReceivedEventDate = messageCreatedAt,
-                unreadMessages = read.unreadMessages + 1,
+                lastReceivedEventDate = receivedEventDate,
+                unreadMessages = read.let {
+                    val hasNewUnreadMessage = receivedEventDate.after(it.lastReceivedEventDate) &&
+                        newMessages.size > messages.size &&
```

Comment:
I assume this condition is always true, since `newMessages` equals to `messages` + `message`?

Reply:
It should, but during the development, I had some cases where it wasn't the case and the counter counts twice